### PR TITLE
Improve OAuth2 error handling

### PR DESCRIPTION
**Pull request by stnguyen90**
## What does this PR do?

There are two problems when the OAuth2 flow fails.

First, if there is a problem during authentication, the user will be redirected back to Appwrite without the authorization code, and Appwrite will throw an error like this:

image

The other problem occurs if Appwrite receives an authorization code but fails to exchange it for an access token. In that case, an error like this occurs:

image

This PR improves both scenarios to show more detailed info or send the data to the failure URL for the developer.

Without a failure URL, the user will see a more detailed error:

image

With a failure URL, the user will be redirected to it like:

```
https://localhost/auth/oauth2/failure?error=%7B%22message%22%3A%22Failed+to+obtain+access+token.+The+Google+OAuth2+provider+returned+an+error%3A+invalid_grant%3A+Malformed+auth+code.%22%2C%22type%22%3A%22user_oauth2_bad_request%22%2C%22code%22%3A400%7D#
```

The developer can then pull the `error` from the query string with something like:

```javascript
u = new URL(window.location);
e = JSON.parse(u.searchParams.get('error'));
```

which would yield:

```json
{
    "message": "Failed to obtain access token. The Google OAuth2 provider returned an error: invalid_grant: Malformed auth code.",
    "type": "user_oauth2_bad_request",
    "code": 400
}
```

## Test Plan

Manual

## Related PRs and Issues

- https://github.com/appwrite/appwrite/issues/2591
- https://github.com/appwrite/appwrite/issues/4753

## Checklist

- [x] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [x] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs?

### CODE REVIEW

File: app/config/errors.php [hunk 1]
```diff
@@ -195,6 +195,21 @@
         'description' => 'Missing ID from OAuth2 provider.',
         'code' => 400,
     ],
+    Exception::USER_OAUTH2_BAD_REQUEST => [
+        'name' => Exception::USER_OAUTH2_BAD_REQUEST,
+        'description' => 'OAuth2 provider rejected the bad request.',
```
Comment (gewenyu99) [inline app/config/errors.php:200]: This is great!

I would still elaborate on these a little:
**[What went wrong] [Why we think it went wrong] [How you could fix it]**

OAuth2 provider rejected the bad request. We think [SOMETHING] is misconfigured. Check [PLACE] to see if there are any mistakes.

**Review by gewenyu99 — CHANGES_REQUESTED** [review]
Love where this is going.

I'd still like to see more context, even if we're being too wordy, or giving educated guesses. More information for the end user to use to debug is always valuable.

File: app/config/errors.php (hunk 1 above)
Comment (stnguyen90) [reply app/config/errors.php:200]: @gewenyu99, these are the default descriptions shown in the docs, but they can be different from what's returned to the developer. For example, one error response we have is:

```json
{
    "message": "Failed to obtain access token. The Google OAuth2 provider returned an error: invalid_grant: Malformed auth code.",
    "type": "user_oauth2_bad_request",
    "code": 400
}
```

Comment (gewenyu99) [reply app/config/errors.php:200]: Okay, would I be so bold to suggest saying: OAuth2 provider rejected the bad request. This error will be replaced by an error from the provider at runtime.

I know it's excessive, but maybe this is a good hint for maintainers, but also in docs.

I err on the side of over explaining with errors cause it's extra frustrating. You can make the call.

**Review by gewenyu99 — APPROVED** [review]

File: app/config/errors.php (hunk 1 above)
Comment (stnguyen90) [reply app/config/errors.php:200]: > This error will be replaced by an error from the provider at runtime.

That doesn't really make sense to me because it's just a description of the error type.

File: app/controllers/api/account.php [hunk 2]
```diff
@@ -447,27 +461,66 @@
...
+        try {
+            $accessToken = $oauth2->getAccessToken($code);
+            $refreshToken = $oauth2->getRefreshToken($code);
+            $accessTokenExpiry = $oauth2->getAccessTokenExpiry($code);
+        } catch (OAuth2Exception $ex) {
+            $failureRedirect(
+                $ex->getType(),
+                'Failed to obtain access token. The ' . $providerName . ' OAuth2 provider returned an error: ' . $ex->getMessage(),
+                $ex->getCode(),
+            );
```
Comment (stnguyen90) [inline app/controllers/api/account.php:518]: @gewenyu99 ,

Here, `$ex->getType()` can either be:

1. `USER_OAUTH2_BAD_REQUEST`
2. 'USER_OAUTH2_UNAUTHORIZED`
3. `USER_OAUTH2_PROVIDER_ERROR`

So, I can put some additional text based on that. I just need to know what you'd want to say.

So far, from my tests, `USER_OAUTH2_BAD_REQUEST` occurs if the code is old or incorrect.

`USER_OAUTH2_UNAUTHORIZED` can occur if the secret in appwrite is incorrect.

**Comment by eldadfux** [comment]
@stnguyen90 also, something has happened to the style of the HTML error page. Let's see if we can replicated the standalone error style we use for the console.

File: app/config/errors.php (hunk 1 above)
Comment (gewenyu99) [reply app/config/errors.php:200]: > > This error will be replaced by an error from the provider at runtime.
> 
> That doesn't really make sense to me because it's just a description of the error type.

:/ Maybe the real issue is how we present these errors. I never know which ones will have their description replaced at runtime, which ones will actually use the description in the error file.

I'll let it slide, cause it's not really developer facing, more a maintainer facing issue.

**Review by gewenyu99 — APPROVED** [review]
I'm okay with the words used in this PR. Can't speak to the code :P

**Review by eldadfux — APPROVED** [review]
